neqo-http3/src/send_message.rs

```diff
@@ -177,8 +182,20 @@ impl SendStream for SendMessage {
         let available = conn
             .stream_avail_send_space(self.stream_id())
             .map_err(|e| Error::map_stream_send_errors(&e.into()))?;
-        if available <= 2 {
+        if available == 0 {
             return Ok(0);
+        } else if available <= 2 {
+            // Rare case, where available send space at most fits the data frame
+            // header (min 2 bytes), but no data. Don't return `Ok(0)`, but
+            // instead use up `available` anyways, with a small but non-zero
+            // data frame, buffering the remainder, thereby signaling to the
+            // QUIC layer that more is needed, and thus ensure to receive
+            // [`neqo_transport::ConnectionEvent::SendStreamWritable`] when more
+            // send space is available.
+            const MIN_DATA: usize = 1; // arbitrary small value, larger zero
+            return self
+                .send_data_atomic(conn, &buf[..MIN_DATA])
+                .map(|()| MIN_DATA);
```

Comment:
I'd like to see a test showing that this is OK.  I'm fairly sure that it isn't.

You aren't sending a DATA frame header for this.  That header is always at least 2 bytes.

Reply:
@martinthomson the DATA frame header is sent as part of the call to `self.send_data_atomic`:

https://github.com/mozilla/neqo/blob/d74ad5ea9b6dbed83674fc1963a6424c79c12588/neqo-http3/src/send_message.rs#L295-L305

Not saying that this doesn't need a test.

Reply:
Ah, so that will result in us sending three bytes to the stream, in a case where we only have space for 2 or less.  How does that ever work?

Reply:
> in a case where we only have space for 2 or less

We will use up the available space and buffer the remaining. E.g. say that we have 2 bytes available, we send the header (2bytes) and buffer 1 byte. 

Once more space is available, we get an event (because we previously exhausted the send space) and can thus first empty the 1byte from the buffer and then send more goodput. 

Does that make sense @martinthomson?

(Again, not particularly happy with this hack. Still thinking about alternatives.)